### PR TITLE
Iteration3 win percentage da

### PR DESCRIPTION
**Pull request by d-atkins**
Add the following in game_teams_collection.rb:
- game_teams_hash: sorts game_teams into a hash making each key a team id and each corresponding value all of the games associated with that team
- percentage: takes a numerator and denominator and returns a percentage (to be refactored into a module later)
- team_win_percentage: calculates the win percentage of all games, or if an optional argument is given, the win percentage of only the home or away games
- winningest_team_id: finds all teams' win percentages and returns the team with the highest rate
- hoa_differences: finds the difference of home win rate % minus away win rate % for each team
- best_fans: returns the team with the highest positive hoa diff
- worst_fans: returns an array with all teams with a negative hoa diff

### CODE REVIEW

File: lib/game_teams_collection.rb [hunk 1]
```diff
@@ -13,21 +13,66 @@ def initialize(file_path)
...
+  end
 
-  def game_teams_by_id
+  def percentage(numerator, denominator) #to-do: make Calculatable module
+    return ((numerator.to_f / denominator) * 100).round(2)
+  end
 
-    hash.keys.reduce({}) do |new_hash, key|
-      new_hash[key] = hash[key].find_home_games
+  def team_win_percentage(games_hash, team, hoa = nil)
```
Comment (mronauli) [inline lib/game_teams_collection.rb:30]: nice hoa = nil, seems robust/efficient

File: lib/game_teams_collection.rb [hunk 2]
```diff
@@ -13,21 +13,66 @@ def initialize(file_path)
...
+  def game_teams_hash
+    @game_teams_array.reduce({}) do |hash, game_teams|
+      hash[game_teams.team_id] << game_teams if hash[game_teams.team_id]
+      hash[game_teams.team_id] = [game_teams] if hash[game_teams.team_id].nil?
+      hash
+    end
+  end
 
-  def game_teams_by_id
+  def percentage(numerator, denominator) #to-do: make Calculatable module
```
Comment (tylertomlinson) [inline lib/game_teams_collection.rb:26]: I like this. When we refactor more could also build average methods.

File: lib/game_teams_collection.rb [hunk 3]
```diff
@@ -13,21 +13,66 @@ def initialize(file_path)
...
+    hash.each do |key, value|
+      win_percentages[key] = team_win_percentage(hash, key)
+    end
+    win_percentages.max_by { |key, value| value }[0]
   end
 
-  def find_home_games
-    game_teams_lists_by_id[team_id].find_all do |game_teams|
-      game_teams.hoa == "home"
+  def hoa_differences(hash)
```
Comment (mronauli) [inline lib/game_teams_collection.rb]: could be more explicit in terms of where hash argument is coming from

Comment (tylertomlinson) [inline lib/game_teams_collection.rb]: Nice method

File: lib/game_teams_collection.rb (hunk 2 above)
Comment (d-atkins) [reply lib/game_teams_collection.rb:26]: yes

File: lib/game_teams_collection.rb (hunk 3 above)
Comment (d-atkins) [reply lib/game_teams_collection.rb]: noice

File: lib/game_teams_collection.rb [hunk 4]
```diff
@@ -13,21 +13,66 @@ def initialize(file_path)
   def create_game_teams_array(file_path)
     load_from_csv(file_path, GameTeams)
   end
-  # require "pry"; binding.pry
```
Comment (tylertomlinson) [inline lib/game_teams_collection.rb:16]: Great work taking out pry everywhere!

File: test/game_teams_collection_test.rb [hunk 5]
```diff
@@ -20,4 +20,36 @@ def test_it_can_create_games_from_csv
...
+  end
+
+  def test_it_can_find_id_of_team_with_best_fans
+    assert_equal "20", @game_teams_collection.best_fans_id
+  end
+
+  def test_it_can_find_teams_with_worst_fans
+    expected = ["16", "14", "5", "28", "26", "19"]
+    assert_equal expected, @game_teams_collection.worst_fans_ids
+  end
```
Comment (mronauli) [inline test/game_teams_collection_test.rb:54]: it all looks good!

File: lib/game_teams_collection.rb [hunk 6]
```diff
@@ -13,4 +13,12 @@ def create_game_teams_array(file_path)
 
     csv.map { |row| GameTeams.new(row) }
   end
+
+  def game_teams_hash
```
Comment (Yetidancer) [inline lib/game_teams_collection.rb:18]: can be added as a instance variable to increase performance

File: test/game_teams_collection_test.rb (hunk 5 above)
Comment (d-atkins) [reply test/game_teams_collection_test.rb:54]: good

File: lib/game_teams_collection.rb (hunk 6 above)
Comment (d-atkins) [reply lib/game_teams_collection.rb:18]: yes!

File: lib/game_teams_collection.rb (hunk 3 above)
Comment (d-atkins) [reply lib/game_teams_collection.rb]: renamed to all_games

**Pull request merged**